Finished first draft of Updates to workflows (#113)
griptape-ai/griptape-trade-school: docs/courses/compare-movies-workflow/08_updates_imperative_vs_desclarative.md

# Updates to Workflows

> ## Imperative vs Declarative Syntax

Reviewer: I wonder if it'd be good to write a quick reminder of what Imperative vs Declarative means?

example:

With Imperative declaration - you defining a specific series of tasks and then define their relationships.
With Declarative, we are defining the goal - the tasks are described in terms of their relationships instead of specifcally calling the steps to create them.

or something like that.

> In this course we currently haven't touched on the updates to creating Workflows. With some new feautures we've added the ability to specifically declare parent - child relationships. Let's go over the updates! 

Reviewer: It would be nice to mention when this change to workflows happened. Something like: So far this course has been focused on how you could create Workflows with Griptape before v0.27.0. However now it's possible to use a more intuitive syntax for managing parent/child relationships with workflow tasks. In this section we'll take a look at these changes.

or something like that.

!!! note 
    Here is a [link](https://docs.griptape.ai/latest/griptape-framework/structures/workflows/#declarative-vs-imperative-syntax){target="_blank"} to the documentation for Workflows. This page will explain further what we are able to do with these new features.

> Create a new python file and add this code.

Reviewer: Explain that this code demonstrates how we've been creating workflows throughout the course - by creating a start and end task, then inserting tasks between them.

```python 
from dotenv import load_dotenv

> # Griptape

Reviewer: When doing code snippets, try adding some spacing before comments to help the flow read better.

For example:

```
from dotenv import load_dotenv

# Griptape
from griptape.tasks import PromptTask
from griptape.structures import Workflow

# Load env variables
load_dotenv()
```

from griptape.tasks import PromptTask
from griptape.structures import Workflow
# Load env variables
load_dotenv()
# Create the workflow
workflow = Workflow()
# Create the start task
start_task = PromptTask("This is the start", id="START")
# Create the end task, and pass it the previous task's results
end_task = PromptTask("What did we do? {{ parent_outputs }}", id="END")
# Add the tasks to the workflow
workflow.add_tasks(start_task, end_task)
# List of movies
movies = ["The Matrix", "Toy Story"] 
# Empty tasks list 
tasks = []
for movie in movies:
    # Creates a task for every movie
    task = PromptTask(
        f"Create a prompt for the movie {movie}",
        context={"movie": movie},
        # Dynamically set id based on movie name.
        id=f"{movie}_task"
    )
    # Append the tasks to the list
    tasks.append(task)
    # Insert the tasks into the workflow
workflow.insert_tasks(start_task, [*tasks], end_task)

workflow.run()
```
While the above code gets us the output we desired, there are better ways to create workflows. Let's dive into those! 

## A better way to build Workflows

> Let's change around a few lines!  

Reviewer: Explain what we're going to do here - why are we changing lines? what will we be doing?

A good rule of thumb when creating tutorials or doing demos is this:

1) Tell them what you're going to do ("Instead of _inserting_ tasks, we're going to explicitly define a parent or child.")
2) Do the thing (Show the code)
3) Tell them what you did ("Notice how we used the `add_parent` command to add a particular task as a parent." )

That rule of 3 is a great way to make sure the ideas stick in their heads.

```python hl_lines="14-15"
# Load env variables
load_dotenv()

start_task = PromptTask("This is the start", id="START")
end_task = PromptTask("What did we do? {{ parent_outputs }}", id="END")
movies = ["StarWars Episode V: The Empire Strikes Back", "Dune 2"]
tasks = []
for movie in movies:
    task = PromptTask(
        f"Create a prompt for the movie {movie}",
        context={"movie": movie},
        id=f"TASK_{movie}"
    )
    task.add_parent(start_task)
    end_task.add_parent(task)
    tasks.append(task)

> workflow = Workflow(tasks=[start_task, *tasks, end_task])

Reviewer: Highlight this line, because declaring the tasks explicitly as part of the Workflow is new.

Reviewer: We might also want to explain what `*tasks` is doing - this is an intermediate/advanced concept that is super useful to understand, but may not be obvious.

workflow.run()
```

> As you can see we added a new way to declare Parent-child relationships, more on this later. Let's look at another way we can write this exact same workflow. 

Reviewer: Call out that in this case, we're going to declare the parent & child ids _while_ creating the task.

```python hl_lines="9-10"
load_dotenv()
movies=["StarWars Episode V: The Empire Strikes Back", "Dune 2"]
tasks = []
for movie in movies:
    task = PromptTask(
        f"Create a prompt for the movie {movie}",
        context={"movie": movie}, 
        id=f"TASK_{movie}",
        parent_ids=["START"],
        child_ids=["END"]        
    )
    tasks.append(task)
workflow = Workflow(
    tasks = [
        PromptTask("This is the start", id="START"),
        *tasks,
        PromptTask("What did we do?", id="END")
    ]

> )

Reviewer: Highlight this section as well, because needing to declare what tasks are part of the workflow is new.

workflow.run()
```

> These two workflows above output exactly the same, the only thing that is different is what syntax we use to define Parent-Child relationships. More on this later. This comes in handy when dealing with larger workflows. Below we show the changes top to bottom.

Reviewer: ```suggestion
These two workflows create the same output, the only thing that is different is what syntax we use to define Parent-Child relationships. More on this later. This comes in handy when dealing with larger workflows. Below we show the changes top to bottom.
```

> 

Reviewer: This section is a great summary - maybe give it a header "## Summary" or something?

In Griptape 0.26 this is most likely how your code would've looked: 
```python
workflow = Workflow()

workflow.add_tasks(start_task, end_task)
 
workflow.insert_tasks(start_task, [*tasks], end_task)
```
     
Now in Griptape 0.27 the syntax for creating workflows is much lighter, and leads to a better dev experience.
```python
workflow = Workflow(tasks=[start_task, *tasks, end_task])
```
Let's move on and look at the updates to declaring parent-child relationships!

## Updates to Parent-Child relationships

> Previously in Griptape 0.26 declaring parent-child relationships was pretty confusing. You first would have to add the start and end tasks, then all children tasks were inserted between the parent and last child. With Griptape 0.27 this process not only easier to visualize, but also easier to use. Let's take a look at imperatively declaring parent-children relationships.

Reviewer: Instead of saying it was "pretty confusing", say "wasn't always very intuitive". :)

### Imperatively specifying Parent-Child relationships(same as above):
```python hl_lines="10-11"
start_task = PromptTask("This is the start")
end_task = PromptTask("What did we do? {{ parent_outputs }}")
movies = ["The Matrix", "Toy Story"] #The movies
tasks = []
for movie in movies:
    task = PromptTask(
        f"Create a prompt for the movie {movie}",
        context={"movie": movie},
    )

>     task.add_parent(start_task)

Reviewer: Add a comment:

# Specifically add the start_task as a parent to the current task.

    end_task.add_parent(task)
    tasks.append(task)

> workflow = Workflow(tasks=[start_task, *tasks, end_task])

Reviewer: Add a newline before the workflow line - to help readability:

```
    end_task.add_parent(task)
    tasks.append(task)

# Create the workflow, and specify the tasks
workflow = Workflow(tasks=[start_task, *tasks, end_task])
```

workflow.run()
```

> If you wanted to you could also specify children, instead of parent relationships like this:

Reviewer: Great call out!

```python 
start_task = PromptTask("This is the start", id="START")
end_task = PromptTask("What did we do? {{ parent_outputs }}", id="END")
movies = ["The Matrix", "Toy Story"] #The movies
tasks = []
for movie in movies:
    task = PromptTask(
        f"Create a prompt for the movie {movie}",
        context={"movie": movie},
        id=f"TASK_{movie}"
    )

>     start_task.add_child(task)

Reviewer: Highlight this section, and point out we're calling the children.

    task.add_child(end_task)
    tasks.append(task)
workflow = Workflow(tasks=[start_task, *tasks, end_task])
workflow.run()
``` 
These both output the following graph: 
``` mermaid
graph TB 
    subgraph " "
        direction TB
        B(start_task):::main --> C(TASK_Toy story) --> E(END):::main
        B --> D(TASK_The Matrix) --> E
    end
    
    classDef main fill:#4274ff1a, stroke:#426eff
```

> If you added a few more movies the graph would be: 

Reviewer: Nice!

``` mermaid
graph TB 
    subgraph " "
        direction TB
        B(start_task):::main --> C(TASK_Toy story) --> E(END):::main
        B --> D(TASK_The Matrix) --> E
        B --> F(TASK_Spiderman) --> E
        B --> G(TASK_Spiderman 2) --> E
    end
    
    classDef main fill:#4274ff1a, stroke:#426eff
```

> Now that we've seen how we can imperatively declare Parent-Child relationships, let's learn how to do it declaratively! I've changed around the example in order to better explain how this works.

Reviewer: Add a reminder of what Delcarative means.

## Declaratively specificying Parent-Child relationships 
Declaratively specifying Parent relationships:
```python
load_dotenv()

workflow = Workflow(
    tasks = [
        PromptTask(
            "Generate two random movies from the 90's", 
            id="START"
            ),
        PromptTask(
            "Given two movies, summarize them: {{parent_outputs}}",
             id="SUMMARIZE",

>              parent_ids = ["START"]

Reviewer: Highlight these lines

also, explain why this is a list of items instead of a single item.

            id="RANK_END"
        )
    ]
)
``` 
Or even a mix of specifying Parent-Child relationships:
```python
workflow = Workflow(
    tasks = [
        PromptTask(
            "Generate two random movies from the 90's", 
            id="START"
            ),
        PromptTask(
            "Given two movies, summarize them: {{parent_outputs}}", 
            id="SUMMARIZE", 
            parent_ids=["START"] ,
            child_ids=["RANK_END"]
            ),
        PromptTask(
            "Compare the two movies: {{parent_outputs}}", 
            id="COMPARE", 
            child_ids=["RANK_END"], 
            parent_ids=["START"]
            ),
        PromptTask(
            "Rank the movies 1 and 2 {{parent_outputs}}}",
            id="RANK_END"
            )
    ]
)
```
All of the above code shares the same workflow structure, see the below graph: 
```mermaid
    graph TB 
    subgraph " "
        direction TB
        B(start_task):::main --> C(COMPARE) --> E(RANK_END):::main
        B --> D(SUMMARIZE) --> E
    end
    
    classDef main fill:#4274ff1a, stroke:#426eff
``` 
## Example of specifying both Imperative and Declarative syntax

> Start by emptying your file, all that should be left is the import statements and the load_dotenv().

Reviewer: Why might someone want to do both Imeperative and Declarative? It's cool you can -why would they want to?

```python
from dotenv import load_dotenv
# Griptape
from griptape.structures import Workflow
from griptape.tasks import PromptTask

load_dotenv()
```
Let's get started by creating some tasks! 
```python
load_dotenv()
# Start task, Generates two movies from the 90's
start_task = PromptTask(
    "Generate two movies from the 90's",
    id="START"
)
# Summary task, Summarizes the previously generated movies
summary_task = PromptTask(
    "Summarize these movies: {{parent_outputs}}",
    id="SUMMARIZE"
)
# Compare task, Compares the two movies.
compare_task = PromptTask(
    "Compare these movies: {{parent_outputs}}",
    id="COMPARE"
)
# Takes in the summary and compare tasks to rank the movies
rank_end_task = PromptTask(
    "Rank the movies 1 or 2: {{parent_outputs}}",
    id="RANK_END"
)
```
Now that we have the tasks, we need to create the workflow.
```python
# Under the rank_end_task
workflow = Workflow(
    tasks = [start_task, summary_task, compare_task, rank_end_task]
)
workflow.run()
```

> If you chose to run this code now it wouldn't generate the desired output. That's because we haven't declared the Parent-Children relationships. Let's do that now! 

Reviewer: nice - I like showing how it doesn't work, then showing how to fix it.

```python hl_lines="19-20 22-24"
start_task = PromptTask(
    "Generate two movies from the 90's",
    id="START"
)
# Summary task, Summarizes the previously generated movies
summary_task = PromptTask(
    "Summarize these movies: {{parent_outputs}}",
    id="SUMMARIZE"
)
# Compare task, Compares the two movies.
compare_task = PromptTask(
    "Compare these movies: {{parent_outputs}}",
    id="COMPARE"
)
# Takes in the summary and compare tasks to rank the movies
rank_end_task = PromptTask(
    "Rank the movies 1 or 2: {{parent_outputs}}",
    id="RANK_END",
    # Declaratively specifying Parents
    parent_ids=["COMPARE", "SUMMARIZE"]
)

> # Imperatively setting the parents

Reviewer: Great! just add some line numbers between sections so it's easier to read.

```
start_task = PromptTask(
    "Generate two movies from the 90's",
    id="START"
)

# Summary task, Summarizes the previously generated movies
summary_task = PromptTask(
    "Summarize these movies: {{parent_outputs}}",
    id="SUMMARIZE"
)

# Compare task, Compares the two movies.
compare_task = PromptTask(
    "Compare these movies: {{parent_outputs}}",
    id="COMPARE"
)

# Takes in the summary and compare tasks to rank the movies
rank_end_task = PromptTask(
    "Rank the movies 1 or 2: {{parent_outputs}}",
    id="RANK_END",
    # Declaratively specifying Parents
    parent_ids=["COMPARE", "SUMMARIZE"]
)

# Imperatively setting the parents
summary_task.add_parent(start_task)
compare_task.add_parent(start_task)
```

summary_task.add_parent(start_task)
compare_task.add_parent(start_task)
```
!!!note
    You can now place tasks in whatever order you see fit within the tasks list inside of the workflow strucure call.

> Now that we have set the Parent-Child relationships, run the workflow, and see what you get. The responses are very interesting. 

Reviewer: No need for a comma after "run the workflow"

## Code Checkpoint 
This is what your code should look like currently.

> ```python

Reviewer: Go ahead and add line numbers for these code summaries..

you can add `linenums="1"` after ```python

```
This will add the `start_task` as a parent to `my_task`.
The second way is to declare the relationship while creating the Task. For example:
```python
my_task = PromptTask("Do a fun thing", parent_ids=["start_task"])
```
Here, we've used the `parent_ids` parameter to tell the task the `id` of the task that will be it's parent. Notice it's a list - you can have multiple parents for any task. The same is true of children.
Finally, once you've defined the relationships, you will need to add the tasks to the Workflow using the tasks parameter.
For example: 
```python
workflow = Workflow(tasks=[start_task, my_task, my_other_task, end_task])
```
We'll get into greater detail on all these updates in the following sections. Read along to learn more.

## Updates to Parent-Child relationships
Previously in Griptape v0.26.0, declaring parent-child relationships wasn't always very intuitive. First, you would’ve had to add the start and end Tasks, and then insert all children Tasks between the parent and the last child. With Griptape v0.27.0, this process is not only easier to visualize, but also easier to use. Let's take a look at imperatively declaring parent-children relationships.

> ### Imperatively specifying Parent-Child relationships(same as above):

Reviewer: ```suggestion
### Imperatively specifying Parent-Child relationships:
```

```python hl_lines="23-24" linenums="1" title="app.py"
# Load env variables
load_dotenv()

start_task = PromptTask("This is the start", id="START")

end_task = PromptTask("What did we do? {{ parent_outputs }}", id="END")

movies = ["StarWars Episode V: The Empire Strikes Back", "Dune 2"]

# list to place Tasks into
tasks = []

# For loop that loops through the movies list
for movie in movies:
   # Creates the Task for every movie
   task = PromptTask(
       f"Create a prompt for the movie {movie}",
       context={"movie": movie},
       id=f"TASK_{movie}"
   )

> # Specifically add the start_task as a parent to the current Task.

Reviewer: Tab in the comments so they're aligned with the code

```suggestion
    # Specifically add the start_task as a parent to the current Task.
    task.add_parent(start_task)
```

   task.add_parent(start_task)
   end_task.add_parent(task)

# Add the Task to the Tasks array for every movie.
   tasks.append(task)

# Create the Workflow and add the Tasks
workflow = Workflow(tasks=[start_task, *tasks, end_task])

# Run the Workflow
workflow.run()
```
You can specify children instead of parent relationships as well. That would look something like this:
```python hl_lines="23-24" linenums="1" title="app.py"
# Load env variables
load_dotenv()

start_task = PromptTask("This is the start", id="START")

end_task = PromptTask("What did we do? {{ parent_outputs }}", id="END")

movies = ["StarWars Episode V: The Empire Strikes Back", "Dune 2"]

# list to place Tasks into
tasks = []

# For loop that loops through the movies list
for movie in movies:
   # Creates the Task for every movie
   task = PromptTask(
       f"Create a prompt for the movie {movie}",
       context={"movie": movie},
       id=f"TASK_{movie}"
   )

> # Specifically add the task as a child to the start_task
>    start_task.add_child(task)
>    task.add_child(end_task)

Reviewer: Align the comment

```suggestion
    # Specifically add the task as a child to the start_task
    start_task.add_child(task)
    task.add_child(end_task)
```
